### PR TITLE
[SHRINKRES-351] enh: using maven 4 project-local dependency resolution, if available

### LINKED ISSUE #330: [Feature] Resolver for Maven 4
Maven 4 is now in late beta.
I would like to have a Maven 4-compatible version of resolver, in addition to 3

Major issue is (and always has been) that multi-module projects do not work well with resolver.
When executing maven from root, it all works, but executing from within a module leaves
dependencies not found.
With maven 4, this problem can be fixed by using project-local repository.

Thank you!

### PR DESCRIPTION
**Pull request by lprimak**
Uses Maven 4 project-local repository to resolve articles, if available
Compatible with both maven 3 and maven 4
Maven 4 is not required and no dependencies are changed

Added 2 system properties to control new features (described in the README):
- `org.jboss.shrinkwrap.resolver.maven.skipCompilation`: Flag to skip compilation of resolved artifacts (true/false) - default is false.
- `org.jboss.shrinkwrap.resolver.maven.disableProjectLocal`: Flag to disable Maven 4 project-local repository (true/false) - default is false.

fixes #330

### CODE REVIEW

**Comment by lprimak** [comment]
@petrberan Simplest version ready to go!
Thank you

**Comment by lprimak** [comment]
@petrberan Ready to go!

**Comment by lprimak** [comment]
ping :)

**Comment by petrberan** [comment]
Hi @lprimak , sorry about the wait, I was preoccupied with a different issue and it's just now that I can have a look at your PRs. 

I wasn't actually able to find anything in Maven 4 related to project local dependencies and its changes against Maven 3. Do you happen to have some documentation for that?

File: maven/impl-maven-embedded/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/FileExtractor.java [hunk 1]
```diff
@@ -85,8 +85,9 @@ private void extractFileInDestinationDir() {
       } catch (IOException e) {
          System.err.println("Failed to unzip file: " + e.getMessage());
       }
-      markerFileHandler.deleteMarkerFile();
-      System.out.printf("Resolver: Successfully extracted maven binaries from %s%n", fileToExtract);
+      if (markerFileHandler.deleteMarkerFile()) {
```
Comment (petrberan) [inline maven/impl-maven-embedded/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/FileExtractor.java:88]: Even though this is just a simple print, I wouldn't make printing the line dependent on deleting the file. Especially if the file is not related to the actual extraction

File: maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/pom/equipped/PomEquippedEmbeddedMavenRunningAsDaemonTestCase.java [hunk 2]
```diff
@@ -35,7 +35,7 @@ void testDaemonShouldWaitForBuildSuccess() throws TimeoutException {
             .withWaitUntilOutputLineMatches(".*BUILD SUCCESS.*")
             .build();
 
-        Awaitility.await("Wait till thread is not be alive").atMost(20, TimeUnit.SECONDS)
+        Awaitility.await("Wait till thread is not be alive").atMost(45, TimeUnit.SECONDS)
```
Comment (petrberan) [inline maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/pom/equipped/PomEquippedEmbeddedMavenRunningAsDaemonTestCase.java:38]: Why increasing the timeout? Did the changes in the PR resulted in longer run?

File: maven/impl-maven-embedded/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/FileExtractor.java (hunk 1 above)
Comment (lprimak) [reply maven/impl-maven-embedded/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/FileExtractor.java:88]: This is because I was trying to fix flaky test that relies on this message.
Failing to delete marker file indicates a failure in the code, at least that's how I am reading this,
so actual functionality should be correct.

Comment (petrberan) [reply maven/impl-maven-embedded/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/FileExtractor.java:88]: Got it, found the test. Must admit, not my favorite way of checking whether the file was extracted, but I can understand this change.

Can I ask you to put this in a separate commit? As this isn't related to the project-local dependencies I'd prefer git blame to show a related commit when it comes to checking history of this change

File: maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/pom/equipped/PomEquippedEmbeddedMavenRunningAsDaemonTestCase.java (hunk 2 above)
Comment (petrberan) [reply maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/pom/equipped/PomEquippedEmbeddedMavenRunningAsDaemonTestCase.java:38]: Sounds good to me. Can I ask for a separate commit here as well?

**Comment by petrberan** [comment]
Managed to have a look and the code looks solid, thank you @lprimak .

One thing I'd like to discuss here is the introduction of `org.jboss.shrinkwrap.resolver.maven.disableProjectLocal`. I understand that this is something that comes from Maven 4, which is not directly supported in this PR.

The thing that doesn't sit well with me is that this is enabled by default and this could lead to potential backward compatibility issues, as opposed if this was disabled by default and enabled once we support Maven 4.

WDYT?

File: maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/pom/equipped/PomEquippedEmbeddedMavenRunningAsDaemonTestCase.java (hunk 2 above)
Comment (lprimak) [reply maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/pom/equipped/PomEquippedEmbeddedMavenRunningAsDaemonTestCase.java:38]: Yes, I will do that. thanks

File: maven/impl-maven-embedded/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/FileExtractor.java (hunk 1 above)
Comment (lprimak) [reply maven/impl-maven-embedded/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/FileExtractor.java:88]: yes

**Comment by lprimak** [comment]
> Managed to have a look and the code looks solid

Thank you!

> this is enabled by default and this could lead to potential backward compatibility issues

So there is a lot of thought and risk mitigation that went into this.
TLDR: The backwards compatibility risk is minimal and it works out-of-the-box for both maven 3 and 4.

Basically, there are two opposing forces at work here, "works out of the box" and "compatibility"
I evaluated the risk of both. The new functionality will only be invoked if there is `project-local-repo` directory.
This directory is only created by maven 4, so compatibility is pretty much a guarantee here.

Typically maven version is chosen by the users' environment. Both maven 3 and 4 users may work with the same project. What that means is that per-project configuration of the resolver is impractical.
This drives the decision closer to the one I made.

The only "issue" I can see if there is a mixed directory structure, and somehow `project-local-repo` directory of an unrelated project lies as a parent to the current maven project, *and* it has the same required dependency with the same version *and* it's somehow wrong. This is quite a remote possibility, but the alternative is resolver failing, so neither option works for the user anyway.

So, there is only a minuscule risk, and in this case I like to introduce a way to turn off new functionality, which is exactly what I did here.
Just in case I missed some obscure use case :) as we all do.

**Comment by petrberan** [comment]
Understood, thank you for the explanation @lprimak . In the end, we can always file an issue and fix it afterwards, so I don't think there's anything preventing us from merging this.

I'll merge this and release a 3.3.2 next week if you don't mind as I'm not a fan of releasing on Friday, if that's fine :)

Do you want to pursue https://github.com/shrinkwrap/resolver/pull/340 any further or can that be closed?

**Review by petrberan — APPROVED** [review]

**Comment by lprimak** [comment]
Thank you!
 I will rebase #340 once this is merged. Amd then it’ll be ready to go as well

**Comment by petrberan** [comment]
Merging, thank you @lprimak

File: maven/impl-maven/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/MavenWorkingSessionImpl.java [hunk 3]
```diff
@@ -188,6 +197,92 @@ public MavenWorkingSession loadPomFromFile(File pomFile, Properties userProperti
...
+                                .setProperties(Collections.EMPTY_MAP)
+                                .equals(result.getArtifact().setFile(null))))
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * @return absolute path to the project-local repository or null if not found
+     */
+    private Path findProjectLocalRepository() {
+        Path targetPath = Paths.get("target/project-local-repo");
```
Comment (cstamas) [inline maven/impl-maven/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/MavenWorkingSessionImpl.java:274]: What happens if user modifies project output? This is -1

File: maven/impl-maven/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/MavenWorkingSessionImpl.java [hunk 4]
```diff
@@ -188,6 +197,92 @@ public MavenWorkingSession loadPomFromFile(File pomFile, Properties userProperti
...
+
+    private static String toVersionedArtifact(String artifactId, String version) {
+        return artifactId + "-" + version;
+    }
+
+    private void processAdditionalDependencies(Path projectLocalRepository, String groupId,
+                                               String artifactId, String version,
+                                               Set<MavenDependency> additionalDependencies) {
+        Path directory = projectLocalRepository.resolve(groupId).resolve(artifactId).resolve(version);
+        File consumerPom = directory.resolve(toVersionedArtifact(artifactId, version) + "-consumer.pom").toFile();
```
Comment (cstamas) [inline maven/impl-maven/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/MavenWorkingSessionImpl.java:247]: This is Maven internal thing, what happens if layout or naming is modified? This is -1

**Comment by cstamas** [comment]
This PR violates one of the fundamental rules introduced in Maven3 (as Maven2 was a wild-west, in a sense that every plugin, every extension, and even Maven itself had copy paste all around "path operations" directly reaching to local repository). Since then, it was documented:

https://maven.apache.org/repositories/local.html
They key: "While the local repository does reside on local filesystem, users should never reach for its contents directly using plain file operations, but use the provided API instead."

Maven3.9 introduced a TON of warnings, as doing things like these prevented proper adoption of split local repository and other features, that simple broke all code containing assumptions about layout, location and everything.

**Comment by cstamas** [comment]
For context: https://issues.apache.org/jira/browse/MNG-8464

**Comment by lprimak** [comment]
I was told directly by @gnodet 6 months ago that there is no API for project-local repository and I should go directly to the filesystem. I don't think there even an API today to do what this needs to do, so it's really not possible to do what you are suggesting.
